### DISCUSSION TITLE
Support regular expressions in substring

### DISCUSSION THREAD

**Discussion by schallm** [datalust/seq-tickets#1910] [Ideas]
Support substring with a regular expressions and capture group so so it is possible to "capture" text from the middle of a string. For example, how do you to get just the milliseconds out of the following string:

```
The process (processId: 23) took 6.41ms to complete.
```

Postgres:

```sql
substring(Message from 'took ([\d\.]+)ms')  -- returns '6.41'
```

Could something like the following syntax be added?

```sql
substring(Message, /took ([\d\.]+)ms/) -- returns '6.41'
```

Initially brought up as part of [Issue 1386](https://github.com/datalust/seq-tickets/issues/1386).

**Comment by nblumhardt** [comment]
Thanks for the bump on this one @schallm! Good to be reminded of it - still nothing happening in this space in the immediate future, but it intersects with some general string handling work we have on the horizon so I'll make a link in our internal notes and revisit this then 👍

**Reply by schallm** [reply]
That is great. Thanks for the response.

**Comment by schallm** [comment]
A similar usage case we currently use is trying to group our exceptions so we can count them to get common issues.  We have found with our own data; it works well to replace digits and well-known formats with "*" and take the left X characters to group exceptions.  In Postgres this is available using the `regexp_replace` function.  I think we could use the @MessageTemplate to group messages in Seq but would still need something for exceptions.

> This replaces all numbers, hex values 24 characters long, guids, and urls with "*".

```sql
select
  projectnm,
  loggernm,
  loglevel,
  left(regexp_replace(Events.Message, '\d+|[\da-f]{{24}}|[\da-f-]{{36}}|(https?:\/\/)[^\/]*', '\1*', 'gi'), 200) message,
  left(regexp_replace(Events.exception, '\d+|[\da-f]{{24}}|[\da-f-]{{36}}|(https?:\/\/)[^\/]*', '\1*', 'gi'), 200) exception,
  count(*) eventCount
from events
where
  ...
group by
  projectnm,
  loggernm,
  loglevel,
  left(regexp_replace(Events.Message, '\d+|[\da-f]{{24}}|[\da-f-]{{36}}|(https?:\/\/)[^\/]*', '\1*', 'gi'), 200),
  left(regexp_replace(Events.exception, '\d+|[\da-f]{{24}}|[\da-f-]{{36}}|(https?:\/\/)[^\/]*', '\1*', 'gi'), 200)
order by
  eventCount desc
```
Seq would be even better as you don't have to duplicate the columns in the `select` and `group by` clauses.
